File: compiler/doc_utils.py
"""Shared helpers for reading compiled wiki pages and compiler state.

Split out of server.py so each "engine" module (email_engine.py,
resources_engine.py, rag_engine.py) can read frontmatter, look up which page
a topic compiled to, and pull entities/concepts off a raw source's state
entry without importing server.py itself (which would be circular, since
server.py is the thing wiring all the engines together as routes).
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

from models import OUTPUT_DIR
from synthesizer import slugify

COMPILER_DIR = Path(__file__).resolve().parent
INDEX_JSON_PATH = COMPILER_DIR / "temp_output" / "index.json"
LINK_RE = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")


def parse_frontmatter(content: str) -> dict[str, Any]:
    if not content.startswith("---"):
        return {}
    parts = content.split("---", 2)
    if len(parts) < 3:
        return {}
    meta: dict[str, Any] = {}
    tags: list[str] = []
    in_tags = False
    for line in parts[1].splitlines():
        if in_tags:
            if line.startswith("  - "):
                tags.append(line[4:].strip().strip('"').strip("'"))
                continue
            if line.startswith("- "):
                tags.append(line[2:].strip().strip('"').strip("'"))
                continue
            in_tags = False
        match = re.match(r"^(\w+):\s*(.+)$", line)
        if match:
            key = match.group(1)
            value = match.group(2).strip()
            if key == "tags" and value in ("[]", ""):
                in_tags = True
                continue
            if (value.startswith('"') and value.endswith('"')) or (
                value.startswith("'") and value.endswith("'")
            ):
                value = value[1:-1]
            meta[key] = value
            if key == "tags":
                in_tags = True
    if tags:
        meta["tags_list"] = tags
    return meta


def strip_frontmatter(content: str) -> str:
    if not content.startswith("---"):
        return content
    parts = content.split("---", 2)
    return parts[2].lstrip("\n") if len(parts) >= 3 else content


def extract_links(markdown_body: str) -> list[dict[str, str]]:
    return [
        {"text": match.group(1), "href": match.group(2)}
        for match in LINK_RE.finditer(markdown_body)
    ]


def normalize_topic(title: str) -> str:
    return re.sub(r'\\(["\'])', r"\1", title).strip()


def load_topic_index() -> dict[str, str]:
    if not INDEX_JSON_PATH.is_file():
        return {}
    data = json.loads(INDEX_JSON_PATH.read_text(encoding="utf-8"))
    topics = data.get("topics", {})
    return topics if isinstance(topics, dict) else {}


def topic_filename(topic_index: dict[str, str], topic: str, docs_dir: Path | None = None) -> str | None:
    docs_dir = docs_dir or OUTPUT_DIR
    if topic in topic_index:
        return topic_index[topic]
    normalized = normalize_topic(topic)
    for key, filename in topic_index.items():
        if normalize_topic(key) == normalized:
            return filename
    slug = slugify(normalized)
    candidate = f"{slug}.md"
    if (docs_dir / candidate).is_file():
        return candidate
    return None


def collect_source_metadata(state_entry: dict) -> dict[str, Any]:
    """Dedupe the topics/entities/concepts a raw source's chunks were tagged with."""
    topics: list[str] = []
    entities: list[dict[str, str]] = []
    concepts: list[dict[str, str]] = []
    seen_topics: set[str] = set()
    seen_entities: set[str] = set()
    seen_concepts: set[str] = set()

    for chunk in state_entry.get("chunks", []):
        for topic in chunk.get("topics") or []:
            normalized = normalize_topic(topic)
            if normalized and normalized not in seen_topics:
                seen_topics.add(normalized)
                topics.append(normalized)
        for entity in chunk.get("entities") or []:
            name = entity.get("name", "")
            if name and name not in seen_entities:
                seen_entities.add(name)
                entities.append(entity)
        for concept in chunk.get("concepts") or []:
            name = concept.get("name", "")
            if name and name not in seen_concepts:
                seen_concepts.add(name)
                concepts.append(concept)

    return {
        "topics": topics,
        "entities": entities,
        "concepts": concepts,
        "chunks": state_entry.get("chunks", []),
    }


def synthesized_pages_for_topics(
    topics: list[str],
    entities: list[dict[str, str]],
    concepts: list[dict[str, str]],
    docs_dir: Path | None = None,
) -> list[dict[str, Any]]:
    """The compiled wiki pages a source's extracted topics ended up on."""
    docs_dir = docs_dir or OUTPUT_DIR
    topic_index = load_topic_index()
    pages: list[dict[str, Any]] = []
    seen_docs: set[str] = set()

    for topic in topics:
        filename = topic_filename(topic_index, topic, docs_dir)
        if not filename or filename in seen_docs:
            continue
        doc_path = docs_dir / filename
        if not doc_path.is_file():
            continue
        seen_docs.add(filename)
        raw = doc_path.read_text(encoding="utf-8")
        meta = parse_frontmatter(raw)
        body = strip_frontmatter(raw)
        tags = meta.get("tags_list") or []
        pages.append(
            {
                "topic": topic,
                "doc_path": filename,
                "title": meta.get("title") or topic,
                "tags": tags,
                "entities": entities,
                "concepts": concepts,
                "body": body,
                "links": extract_links(body),
            }
        )

    return pages


def read_doc_payload(doc_path: Path, docs_dir: Path | None = None) -> dict[str, Any]:
    docs_dir = docs_dir or OUTPUT_DIR
    rel = str(doc_path.relative_to(docs_dir)).replace("\\", "/")
    raw = doc_path.read_text(encoding="utf-8")
    meta = parse_frontmatter(raw)
    body = strip_frontmatter(raw)
    tags = meta.get("tags_list") or []
    return {
        "path": rel,
        "title": meta.get("title") or doc_path.stem.replace("-", " ").title(),
        "id": meta.get("id"),
        "slug": meta.get("slug"),
        "tags": tags,
        "body": body,
        "links": extract_links(body),
    }


def raw_file_status(rel_path: str, current_md5: str, state: dict) -> str:
    entry = state.get("files", {}).get(rel_path)
    if entry and entry.get("md5") == current_md5:
        return "Processed"
    return "Unprocessed"

File: backend/src/lib/docUtils.ts
/**
 * Port of compiler/doc_utils.py -- read compiled wiki pages (frontmatter +
 * body) and cross-reference them against the topic index the Python
 * linker wrote to compiler/temp_output/index.json.
 */
import fs from 'node:fs';
import path from 'node:path';
import { INDEX_JSON, OUTPUT_DIR } from '../paths';

const LINK_RE = /\[([^\]]+)\]\(([^)]+)\)/g;

export function slugify(text: string): string {
  let s = text.toLowerCase().trim();
  s = s.replace(/[^\w\s-]/g, '');
  s = s.replace(/[\s_]+/g, '-');
  return s.replace(/^-+|-+$/g, '').slice(0, 80);
}

export interface FrontmatterMeta {
  [key: string]: string | string[] | undefined;
  tags_list?: string[];
}

export function parseFrontmatter(content: string): FrontmatterMeta {
  if (!content.startsWith('---')) return {};
  const parts = content.split('---');
  // content.split keeps every "---" occurrence; join back everything after
  // the second one, matching Python's str.split('---', 2)[2] semantics.
  if (parts.length < 3) return {};
  const fmBlock = parts[1];

  const meta: FrontmatterMeta = {};
  const tags: string[] = [];
  let inTags = false;

  for (const line of fmBlock.split('\n')) {
    if (inTags) {
      if (line.startsWith('  - ')) {
        tags.push(line.slice(4).trim().replace(/^["']|["']$/g, ''));
        continue;
      }
      if (line.startsWith('- ')) {
        tags.push(line.slice(2).trim().replace(/^["']|["']$/g, ''));
        continue;
      }
      inTags = false;
    }
    const match = line.match(/^(\w+):\s*(.+)$/);
    if (match) {
      const key = match[1];
      let value = match[2].trim();
      if (key === 'tags' && (value === '[]' || value === '')) {
        inTags = true;
        continue;
      }
      if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
      meta[key] = value;
      if (key === 'tags') inTags = true;
    }
  }
  if (tags.length) meta.tags_list = tags;
  return meta;
}

export function stripFrontmatter(content: string): string {
  if (!content.startsWith('---')) return content;
  const first = content.indexOf('---');
  const second = content.indexOf('---', first + 3);
  if (second === -1) return content;
  return content.slice(second + 3).replace(/^\n+/, '');
}

export function extractLinks(body: string): { text: string; href: string }[] {
  const links: { text: string; href: string }[] = [];
  for (const match of body.matchAll(LINK_RE)) {
    links.push({ text: match[1], href: match[2] });
  }
  return links;
}

export function normalizeTopic(title: string): string {
  return title.replace(/\\(["'])/g, '$1').trim();
}

export function loadTopicIndex(): Record<string, string> {
  if (!fs.existsSync(INDEX_JSON)) return {};
  try {
    const data = JSON.parse(fs.readFileSync(INDEX_JSON, 'utf-8'));
    return data.topics && typeof data.topics === 'object' ? data.topics : {};
  } catch {
    return {};
  }
}

export function topicFilename(
  topicIndex: Record<string, string>,
  topic: string,
  docsDir: string = OUTPUT_DIR,
): string | null {
  if (topic in topicIndex) return topicIndex[topic];
  const normalized = normalizeTopic(topic);
  for (const [key, filename] of Object.entries(topicIndex)) {
    if (normalizeTopic(key) === normalized) return filename;
  }
  const candidate = `${slugify(normalized)}.md`;
  if (fs.existsSync(path.join(docsDir, candidate))) return candidate;
  return null;
}

export interface ChunkLike {
  topics?: string[];
  entities?: { name?: string }[];
  concepts?: { name?: string }[];
}

export function collectSourceMetadata(stateEntry: { chunks?: ChunkLike[] }) {
  const topics: string[] = [];
  const entities: { name?: string }[] = [];
  const concepts: { name?: string }[] = [];
  const seenTopics = new Set<string>();
  const seenEntities = new Set<string>();
  const seenConcepts = new Set<string>();

  for (const chunk of stateEntry.chunks ?? []) {
    for (const topic of chunk.topics ?? []) {
      const normalized = normalizeTopic(topic);
      if (normalized && !seenTopics.has(normalized)) {
        seenTopics.add(normalized);
        topics.push(normalized);
      }
    }
    for (const entity of chunk.entities ?? []) {
      const name = entity.name ?? '';
      if (name && !seenEntities.has(name)) {
        seenEntities.add(name);
        entities.push(entity);
      }
    }
    for (const concept of chunk.concepts ?? []) {
      const name = concept.name ?? '';
      if (name && !seenConcepts.has(name)) {
        seenConcepts.add(name);
        concepts.push(concept);
      }
    }
  }

  return { topics, entities, concepts, chunks: stateEntry.chunks ?? [] };
}

export function synthesizedPagesForTopics(
  topics: string[],
  entities: { name?: string }[],
  concepts: { name?: string }[],
  docsDir: string = OUTPUT_DIR,
) {
  const topicIndex = loadTopicIndex();
  const pages: Record<string, unknown>[] = [];
  const seenDocs = new Set<string>();

  for (const topic of topics) {
    const filename = topicFilename(topicIndex, topic, docsDir);
    if (!filename || seenDocs.has(filename)) continue;
    const docPath = path.join(docsDir, filename);
    if (!fs.existsSync(docPath)) continue;
    seenDocs.add(filename);
    const raw = fs.readFileSync(docPath, 'utf-8');
    const meta = parseFrontmatter(raw);
    const body = stripFrontmatter(raw);
    pages.push({
      topic,
      doc_path: filename,
      title: meta.title || topic,
      tags: meta.tags_list || [],
      entities,
      concepts,
      body,
      links: extractLinks(body),
    });
  }
  return pages;
}

export function readDocPayload(docPath: string, docsDir: string = OUTPUT_DIR) {
  const rel = path.relative(docsDir, docPath).split(path.sep).join('/');
  const raw = fs.readFileSync(docPath, 'utf-8');
  const meta = parseFrontmatter(raw);
  const body = stripFrontmatter(raw);
  const stem = path.basename(docPath, path.extname(docPath));
  const titleCase = stem
    .replace(/-/g, ' ')
    .replace(/\w\S*/g, (w) => w.charAt(0).toUpperCase() + w.slice(1));
  return {
    path: rel,
    title: meta.title || titleCase,
    id: meta.id,
    slug: meta.slug,
    tags: meta.tags_list || [],
    body,
    links: extractLinks(body),
  };
}

export function rawFileStatus(
  relPath: string,
  currentMd5: string,
  state: { files?: Record<string, { md5?: string }> },
): 'Processed' | 'Unprocessed' {
  const entry = state.files?.[relPath];
  return entry && entry.md5 === currentMd5 ? 'Processed' : 'Unprocessed';
}
